components/org.wso2.carbon.identity.application.authenticator.oidc/src/main/java/org/wso2/carbon/identity/application/authenticator/oidc/logout/idpinit/processor/FederatedIdpInitLogoutProcessor.java

```
@@ -0,0 +1,618 @@
+     */
+    private LogoutResponse.LogoutResponseBuilder logoutUsingSid(String sid, String userId)
+            throws LogoutServerException {
+
+        if (log.isDebugEnabled()) {
+            log.debug("Trying federated IdP initiated logout using sid: " + sid);
+        }
+        String sessionId = getSessionIdFromSid(sid);
+        if (sessionId == null) {
+            return new LogoutResponse.LogoutResponseBuilder(HttpServletResponse.SC_INTERNAL_SERVER_ERROR,
+                    ErrorMessages.LOGOUT_SERVER_EXCEPTION.getMessage());
+        }
+
+        UserSessionManagementService userSessionManagementService = OpenIDConnectAuthenticatorDataHolder.getInstance()
+                .getUserSessionManagementService();
+        try {
+            userSessionManagementService.terminateSessionBySessionId(userId, sessionId);
+            log.info("Session terminated for session Id: " + sessionId + " and userId: " + userId);
+            return new LogoutResponse.LogoutResponseBuilder(HttpServletResponse.SC_OK,
+                    OIDCAuthenticatorConstants.BackchannelLogout.LOGOUT_SUCCESS);
+        } catch (SessionManagementException e) {
+            throw handleLogoutServerException(
+                    ErrorMessages.FEDERATED_SESSION_TERMINATION_FAILED, sessionId);
+        }
+    }
+
+    private String getSessionIdFromSid(String sid) throws LogoutServerException {
+
+        try {
+            UserSessionDAO userSessionDAO = new UserSessionDAOImpl();
+            FederatedUserSession federatedUserSession = userSessionDAO.getFederatedAuthSessionDetails(sid);
+            if (federatedUserSession == null) {
+                throw handleLogoutServerException(ErrorMessages.RETRIEVING_SESSION_ID_MAPPING_FAILED, sid);
+            }
+            String sessionId = federatedUserSession.getSessionId();
+            if (StringUtils.isBlank(sessionId)) {
+                throw handleLogoutServerException(ErrorMessages.RETRIEVING_SESSION_ID_MAPPING_FAILED, sid);
```

do we need to throw an exception here? This can occur when there is already a logout happened and sessionId corresponding for federatedIdp might have cleared. we can add debug logs and return